### PR TITLE
Allow generic signature superclass constraints to contain type parameters

### PR DESCRIPTION
**Pull request by slavapestov**
@DougGregor, @jckarter Can you take a quick look?

This is NOT for Swift 2.2.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,12 @@ Swift 3
 * Curried function syntax has been removed, and now produces a compile-time
   error.
 
+* Generic signatures can now contain superclass requirements with dependent types,
```
Comment (jckarter) [inline CHANGELOG.md]: "Dependent types" might give people the wrong impression. 'Superclass requirements involving generic parameters' should be clearer.

**Comment by slavapestov** [comment]
@DougGregor You're right, I believe some compiler_crashers regressed because I was overwriting Superclass. I didn't realize it was an instance variable (can we have mandatory 'this' in C++? Just kidding :-) )

**Comment by jckarter** [comment]
Just give your class a dependent template base class to get mandatory `this->`.

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

File: lib/AST/ArchetypeBuilder.cpp [hunk 2]
```diff
@@ -552,10 +552,28 @@ ArchetypeBuilder::PotentialArchetype::getType(ArchetypeBuilder &builder) {
...
 
+  Type superclass;
+
+  if (Superclass) {
+    if (representative->RecursiveSuperclassType) {
+      builder.Diags.diagnose(SuperclassSource->getLoc(),
+                             diag::recursive_superclass_constraint,
+                             Superclass);
+
+      superclass = Type();
```
Comment (DougGregor) [inline lib/AST/ArchetypeBuilder.cpp]: Not actually necessary.

**Comment by slavapestov** [comment]
@swift-ci Please test

**Pull request closed without merging**